api/agent/pure_runner.go

```diff
@@ -27,171 +28,429 @@ import (
+		ch.shutdown(err)
 	}
+	return err
+}
 
-	// Only write if we are not in an error situation. If we cause a stream error, then record that but don't cancel
-	// the call: basically just blackhole the output and return the write error to cause Submit to fail properly.
-	ch.stateMutex.Lock()
-	defer ch.stateMutex.Unlock()
-	err = ch.streamError
+// enqueueCallResponse enqueues a Submit() response to the LB
+// and initiates a graceful shutdown of the session.
+func (ch *callHandle) enqueueCallResponse(err error) {
+	defer ch.finalize()
+
+	// Error response
 	if err != nil {
-		return 0, fmt.Errorf("Bailing out because of communication error: %v", ch.streamError)
+		ch.enqueueMsgStrict(&runner.RunnerMsg{
+			Body: &runner.RunnerMsg_Finished{Finished: &runner.CallFinished{
+				Success: false,
+				Details: fmt.Sprintf("%v", err),
+			}}})
+		return
 	}
 
-	logrus.Debugf("Sending call response data %d bytes long", len(data))
-	err = ch.engagement.Send(&runner.RunnerMsg{
+	// EOF and Success response
+	ch.enqueueMsgStrict(&runner.RunnerMsg{
 		Body: &runner.RunnerMsg_Data{
 			Data: &runner.DataFrame{
-				Data: data,
-				Eof:  false,
+				Eof: true,
 			},
 		},
 	})
+
+	ch.enqueueMsgStrict(&runner.RunnerMsg{
+		Body: &runner.RunnerMsg_Finished{Finished: &runner.CallFinished{
+			Success: true,
+			Details: ch.c.Model().ID,
+		}}})
+}
+
+// enqueueAck enqueues a ACK or NACK response to the LB for ClientMsg_Try
+// request. If NACK, then it also initiates a graceful shutdown of the
+// session.
+func (ch *callHandle) enqueueAck(err error) error {
+	// NACK
 	if err != nil {
-		ch.streamError = err
-		return 0, fmt.Errorf("Error sending data: %v", err)
+		err = ch.enqueueMsgStrict(&runner.RunnerMsg{
+			Body: &runner.RunnerMsg_Acknowledged{Acknowledged: &runner.CallAcknowledged{
+				Committed: false,
+				Details:   fmt.Sprintf("%v", err),
+			}}})
+		if err != nil {
+			return err
+		}
+		return ch.finalize()
 	}
-	return len(data), nil
+
+	// ACK
+	return ch.enqueueMsgStrict(&runner.RunnerMsg{
+		Body: &runner.RunnerMsg_Acknowledged{Acknowledged: &runner.CallAcknowledged{
+			Committed:             true,
+			Details:               ch.c.Model().ID,
+			SlotAllocationLatency: time.Time(ch.allocatedTime).Sub(time.Time(ch.receivedTime)).String(),
+		}}})
 }
 
-func (ch *callHandle) Close() error {
-	err := ch.commitHeaders()
-	if err != nil {
-		return fmt.Errorf("Error sending close frame: %v", err)
+// spawnPipeToFn pumps data to Function via callHandle io.PipeWriter (pipeToFnW)
+// which is fed using input channel.
+func (ch *callHandle) spawnPipeToFn() chan *runner.DataFrame {
+
+	input := make(chan *runner.DataFrame)
+	go func() {
+		defer ch.closePipeToFn()
+		for {
+			select {
+			case <-ch.doneQueue:
+				return
+			case <-ch.ctx.Done():
+				return
+			case data := <-input:
+				if data == nil {
+					return
+				}
+
+				if len(data.Data) > 0 {
+					_, err := io.CopyN(ch.pipeToFnW, bytes.NewReader(data.Data), int64(len(data.Data)))
```

Comment:
it does seem like we could use the data channel now to send the input on to synchronize with the other goroutine that is filling in the call body (ghost reader style, basically, we just need a `**io.Reader` that implements `io.Reader` pretty much. no need to do anything here, future simplification idea tossing